### PR TITLE
sql: Mandate CSR connections use `CONNECTION`

### PR DESCRIPTION
**Pull request by sploiselle**
Remove the ability to define inline connections to confluent schema registries. We have enough of the `WITH` junk removed that this was tractable to do.

### Motivation

This PR adds a known-desirable feature. Fixes https://github.com/MaterializeInc/materialize/issues/14296

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Require all `SOURCE`s and `SINK`s connecting to Confluent Schema Registries to do so using a `CONNECTION`.

### CODE REVIEW

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 1]
```diff
@@ -149,9 +149,8 @@ impl<T: AstInfo> AstDisplay for ProtobufSchema<T> {
 impl_display_t!(ProtobufSchema);
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum CsrConnection<T: AstInfo> {
-    Inline { url: String },
-    Reference { connection: T::ObjectName },
+pub struct CsrConnection<T: AstInfo> {
+    pub connection: T::ObjectName,
```
Comment (guswynn) [inline src/sql-parser/src/ast/defs/ddl.rs:153]: cleannnn

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 2]
```diff
@@ -207,34 +190,17 @@ impl_display_t!(CsrConnectionAvro);
...
-                f.write_str("CONNECTION ");
-                f.write_node(connection);
-            }
-        }
+        f.write_str("USING CONFLUENT SCHEMA REGISTRY CONNECTION");
+        f.write_node(&self.connection.connection);
 
         if let Some(seed) = &self.seed {
             f.write_str(" ");
             f.write_node(seed);
```
Comment (guswynn) [inline src/sql-parser/src/ast/defs/ddl.rs:202]: hmm, i forgot that we have this, when we purify proto sources...pretty spooky

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 3]
```diff
@@ -207,34 +190,17 @@ impl_display_t!(CsrConnectionAvro);
...
-                f.write_str("'");
-                f.write_node(&display::escape_single_quote_string(uri));
-                f.write_str("'");
-            }
-            CsrConnection::Reference { connection, .. } => {
-                f.write_str("CONNECTION ");
-                f.write_node(connection);
-            }
-        }
+        f.write_str("USING CONFLUENT SCHEMA REGISTRY CONNECTION");
```
Comment (guswynn) [inline src/sql-parser/src/ast/defs/ddl.rs:197]: we need a space here

File: src/sql-parser/tests/testdata/ddl [hunk 4]
```diff
@@ -1373,17 +1373,13 @@ CreateConnection(CreateConnectionStatement { name: UnresolvedObjectName([Ident("
...
+reparse failed: Expected CONNECTION, found identifier "connectionconn2"
 
 
 parse-statement
 CREATE SOURCE src1 FROM KAFKA CONNECTION conn1 TOPIC 'baz' FORMAT PROTOBUF USING CONFLUENT SCHEMA REGISTRY CONNECTION conn2 ENVELOPE DEBEZIUM
 ----
-CREATE SOURCE src1 FROM KAFKA CONNECTION conn1 TOPIC 'baz' FORMAT PROTOBUF USING CONFLUENT SCHEMA REGISTRY CONNECTION conn2 ENVELOPE DEBEZIUM
-=>
-CreateSource(CreateSourceStatement { name: UnresolvedObjectName([Ident("src1")]), col_names: [], connection: Kafka(KafkaSourceConnection { connection: Reference { connection: Name(UnresolvedObjectName([Ident("conn1")])), with_options: [] }, topic: "baz", key: None }), legacy_with_options: [], include_metadata: [], format: Bare(Protobuf(Csr { csr_connection: CsrConnectionProtobuf { connection: Reference { connection: Name(UnresolvedObjectName([Ident("conn2")])) }, seed: None, with_options: [] } })), envelope: Some(Debezium(Plain { tx_metadata: [] })), if_not_exists: false, key_constraint: None, with_options: [] })
+reparse failed: Expected CONNECTION, found identifier "connectionconn2"
```
Comment (guswynn) [inline src/sql-parser/tests/testdata/ddl:1382]: reparse failures here are from the missing space in my above comment

File: src/sql-parser/tests/testdata/ddl [hunk 5]
```diff
@@ -1373,17 +1373,13 @@ CreateConnection(CreateConnectionStatement { name: UnresolvedObjectName([Ident("
 parse-statement
 CREATE SOURCE src1 FROM KAFKA CONNECTION conn1 TOPIC 'baz' FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION conn2 ENVELOPE DEBEZIUM
 ----
-CREATE SOURCE src1 FROM KAFKA CONNECTION conn1 TOPIC 'baz' FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION conn2 ENVELOPE DEBEZIUM
-=>
-CreateSource(CreateSourceStatement { name: UnresolvedObjectName([Ident("src1")]), col_names: [], connection: Kafka(KafkaSourceConnection { connection: Reference { connection: Name(UnresolvedObjectName([Ident("conn1")])), with_options: [] }, topic: "baz", key: None }), legacy_with_options: [], include_metadata: [], format: Bare(Avro(Csr { csr_connection: CsrConnectionAvro { connection: Reference { connection: Name(UnresolvedObjectName([Ident("conn2")])) }, key_strategy: None, value_strategy: None, seed: None, with_options: [] } })), envelope: Some(Debezium(Plain { tx_metadata: [] })), if_not_exists: false, key_constraint: None, with_options: [] })
+reparse failed: Expected CONNECTION, found identifier "connectionconn2"
```
Comment (bkirwi) [inline src/sql-parser/tests/testdata/ddl:1376]: For my education: where do the `connectionconn2`s come from here?

File: test/limits/mzcompose.py [hunk 6]
```diff
@@ -465,7 +465,7 @@ def body(cls) -> None:
             print(
                 dedent(
                     f"""
-                     > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KAFKA BROKER '${{testdrive.kafka-addr}}';
+                     > CREATE CONNECTION IF NOT EXISTS kafka_conn FOR KA'FKA BROKER '${{testdrive.kafka-addr}}';
```
Comment (bkirwi) [inline test/limits/mzcompose.py:468]: Typo?

**Review by guswynn — APPROVED** [review]
Looks good! There is one missing space in a `AstDisplay` impl, but otherwise looks good, so ill approve!

This is SO much cleanup omg

File: test/upgrade/check-from-current_source-schema-registry.td [hunk 7]
```diff
@@ -14,4 +14,4 @@
 # injects a `WITH` option into the wrong spot.
 
 > SHOW CREATE SOURCE data
-"materialize.public.data" "CREATE SOURCE \"materialize\".\"public\".\"data\" FROM KAFKA BROKER 'kafka:9092' TOPIC 'testdrive-data-${testdrive.seed}' FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION \"materialize\".\"public\".\"csr_conn\" SEED VALUE SCHEMA '{\"type\":\"record\",\"name\":\"foo\",\"fields\":[{\"name\":\"a\",\"type\":\"int\"}]}' ENVELOPE NONE"
+"materialize.public.data" "CREATE SOURCE \"materialize\".\"public\".\"data\" FROM KAFKA BROKER 'kafka:9092' TOPIC 'testdrive-data-3093084623' FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION\"materialize\".\"public\".\"csr_conn\" SEED VALUE SCHEMA '{\"type\":\"record\",\"name\":\"foo\",\"fields\":[{\"name\":\"a\",\"type\":\"int\"}]}' ENVELOPE NONE"
```
Comment (bkirwi) [inline test/upgrade/check-from-current_source-schema-registry.td:17]: Looks like a bad substitution here?

**Review by bkirwi — COMMENTED** [review]
🔥!!

File: src/sql-parser/tests/testdata/ddl (hunk 5 above)
Comment (sploiselle) [reply src/sql-parser/tests/testdata/ddl:1376]: This is a bad roundtrip; the space after `CONNECTION` is getting deleted, turning it into an identifer, `connectionconn2`.

Comment (bkirwi) [reply src/sql-parser/tests/testdata/ddl]: Makes sense, thanks!

**Pull request merged**